src/interactivity/win32/windowio.cpp

```diff
@@ -167,6 +167,27 @@ void HandleKeyEvent(const HWND hWnd,
         // --- END LOAD BEARING CODE ---
     }
 
+    // - Certain applications like PowerToys and AutoHotKey, due to their keyboard
+    //   remapping feature, send us key events (WM_KEYUP, WM_KEYDOWN, etc.) using
+    //   SendInput(), whose keycode or scancode values could be outside of the
+    //   valid range. We need to filter those events out, as some clients will
+    //   incorrectly translate them to an ascii NULL character.
+    // - If scan code is 0, we can still infer it from the virtual key code as long
+    //   as the virtual key code is valid (0 < vKey < 0xFF).
+    if (VirtualScanCode == 0) [[unlikely]]
+    {
+        // try to infer scancode from virtual keycode
+        auto FullVirtualScanCode = gsl::narrow_cast<WORD>(OneCoreSafeMapVirtualKeyW(VirtualKeyCode, MAPVK_VK_TO_VSC_EX));
+        VirtualScanCode = LOBYTE(FullVirtualScanCode);
+        // if we still don't have a scancode, return.

```

Comment:
I believe the two comments above (179 and 182) are redundant with the code and the large comment on top. The only thing I personally think that might be worth mentioning here (line 182) is that we abuse the fact that `MapVirtualKeyW` returns 0 for invalid vkeys to filter out invalid vkeys.

This however makes me wonder: What if an application sets a non-zero scancode but an invalid vkey? Should we filter those out?

Reply:
In my testing, non-zero `vScan` doesn't generate any NULL character.

Null is only generated when `vScan == 0` and `vKey` is invalid and Character is `\0`.

Reply:
Forgot to mention in the earlier reply, non-zero scan code will not generate Null character regardless of whether virtual keycode is valid or not. So something like `{ ..., vk: 255, vsc: 38, ... }` won't be a problem. That sequence just means that OS couldn't find a valid mapping for scan code to virtual keycode.